<a href="https://colab.research.google.com/github/ssujinK/SEMINAR-IN-adv.-STAT./blob/master/HW3.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **3-1**
### Fashion Mnist 데이터를 잘 적합시키는 CNN 기반 모형 찾기

In [1]:
try:
  # Colab only
  %tensorflow_version 2.x
except Exception:
  pass

TensorFlow 2.x selected.


In [2]:
from __future__ import absolute_import, division, print_function, unicode_literals, unicode_literals

import tensorflow as tf
from tensorflow import keras

import numpy as np
import matplotlib.pyplot as plt

print(tf.__version__)

2.0.0


In [3]:
fashion_mnist = keras.datasets.fashion_mnist

(train_images, train_labels), (test_images, test_labels) = fashion_mnist.load_data()

4423680/4422102 [==============================] - 0s 0us/step


In [0]:
class_names = ['T-shirt/top', 'Trouser', 'Pullover', 'Dress', 'Coat', 'Sandal', 'Shirt', 'Sneaker', 'Bag', 'Ankle boot']

In [5]:
train_images.shape

(60000, 28, 28)

In [6]:
train_labels

array([9, 0, 0, ..., 3, 0, 5], dtype=uint8)

In [7]:
test_images.shape

(10000, 28, 28)

In [0]:
train_images = train_images / 255.0

test_images = test_images / 255.0

In [0]:
train_images = train_images.reshape(train_images.shape[0], 28, 28, 1)
test_images = test_images.reshape(test_images.shape[0], 28, 28, 1)

In [11]:
from tensorflow.keras.layers import Dense, Flatten, Conv2D, MaxPooling2D, AveragePooling2D
from tensorflow.keras import Sequential

model = Sequential()

model.add(Conv2D(input_shape = (28,28,1),filters = 16, kernel_size = (5,5),
                 strides = (1,1), padding='valid', activation ='relu') )
model.add(MaxPooling2D(pool_size = (2,2), strides = (1,1), padding = 'valid') )
model.add(Conv2D(16, kernel_size = (5,5), strides = (1,1), padding='valid', activation ='relu') )
model.add(AveragePooling2D(pool_size = (2,2), strides = (2, 2), padding = 'valid') )
model.add(Conv2D(120, kernel_size=(5, 5), strides=(1, 1), activation='tanh', padding='valid'))
model.add(Flatten())
model.add(Dense(50, activation = 'relu'))
model.add(Dense(10, activation = 'softmax'))

model.summary()

Model: "sequential_1"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
conv2d_1 (Conv2D)            (None, 24, 24, 16)        416       
_________________________________________________________________
max_pooling2d (MaxPooling2D) (None, 23, 23, 16)        0         
_________________________________________________________________
conv2d_2 (Conv2D)            (None, 19, 19, 16)        6416      
_________________________________________________________________
average_pooling2d (AveragePo (None, 9, 9, 16)          0         
_________________________________________________________________
conv2d_3 (Conv2D)            (None, 5, 5, 120)         48120     
_________________________________________________________________
flatten (Flatten)            (None, 3000)              0         
_________________________________________________________________
dense (Dense)                (None, 50)               

In [0]:
model.compile(optimizer = 'adam',
              loss = 'sparse_categorical_crossentropy',
              metrics = ['accuracy'])

In [13]:
model.fit(train_images, train_labels, epochs = 5)

Train on 60000 samples
Epoch 1/5
60000/60000 [==============================] - 115s 2ms/sample - loss: 0.5004 - accuracy: 0.8166
Epoch 2/5
60000/60000 [==============================] - 114s 2ms/sample - loss: 0.3412 - accuracy: 0.8749
Epoch 3/5
60000/60000 [==============================] - 114s 2ms/sample - loss: 0.3058 - accuracy: 0.8873
Epoch 4/5
60000/60000 [==============================] - 114s 2ms/sample - loss: 0.2811 - accuracy: 0.8964
Epoch 5/5
60000/60000 [==============================] - 114s 2ms/sample - loss: 0.2621 - accuracy: 0.9023


In [14]:
test_loss, test_acc = model.evaluate(test_images, test_labels, verbose = 2)

print('\nTest accuracy:', test_acc)

10000/1 - 5s - loss: 0.1928 - accuracy: 0.8925

Test accuracy: 0.8925


Image classifier는 이 데이터셋에서 약 89.3%의 accuracy를 가집니다.

---
# **3-2**
### Mnist 데이터를 잘 적합시키는 CNN 기반 모형 찾기

In [15]:
from __future__ import absolute_import, division, print_function, unicode_literals

!pip install -q tensorflow-gpu==2.0.0-rc1
import tensorflow as tf

     |████████████████████████████████| 380.5MB 43kB/s 
     |████████████████████████████████| 4.3MB 29.2MB/s 
     |████████████████████████████████| 501kB 45.4MB/s 


In [0]:
mnist = tf.keras.datasets.mnist

(x_train, y_train), (x_test, y_test) = mnist.load_data()
x_train, x_test = x_train / 255.0, x_test / 255.0

In [0]:
model = tf.keras.models.Sequential([
  tf.keras.layers.Flatten(input_shape=(28, 28)),
  tf.keras.layers.Dropout(0.2),
  tf.keras.layers.Dense(128, activation='relu'),
  tf.keras.layers.Dense(60, activation='relu'),
  tf.keras.layers.Dropout(0.2),
  tf.keras.layers.Dense(10, activation='softmax')
])


In [0]:
model.compile(optimizer='adam',
              loss='sparse_categorical_crossentropy',
              metrics=['accuracy'])

In [33]:
model.fit(x_train, y_train, epochs=8)

testacc = model.evaluate(x_test,  y_test, verbose=2)
print('\nTest accuracy:', testacc[1])

Train on 60000 samples
Epoch 1/8
60000/60000 [==============================] - 6s 107us/sample - loss: 0.0461 - accuracy: 0.9846
Epoch 2/8
60000/60000 [==============================] - 6s 106us/sample - loss: 0.0451 - accuracy: 0.9853
Epoch 3/8
60000/60000 [==============================] - 6s 107us/sample - loss: 0.0425 - accuracy: 0.9864
Epoch 4/8
60000/60000 [==============================] - 7s 111us/sample - loss: 0.0421 - accuracy: 0.9861
Epoch 5/8
60000/60000 [==============================] - 7s 112us/sample - loss: 0.0392 - accuracy: 0.9868
Epoch 6/8
60000/60000 [==============================] - 6s 106us/sample - loss: 0.0381 - accuracy: 0.9875
Epoch 7/8
60000/60000 [==============================] - 6s 106us/sample - loss: 0.0387 - accuracy: 0.9874
Epoch 8/8
60000/60000 [==============================] - 6s 106us/sample - loss: 0.0353 - accuracy: 0.9881
10000/1 - 0s - loss: 0.0336 - accuracy: 0.9838

Test accuracy: [0.06692937038806067, 0.9838]


In [34]:
print( testacc[1])

0.9838


Classifier는 이 데이터셋에서 약 98.4%의 accuracy를 가집니다.